### ISSUE TITLE
Support variable paging (aka chunked arrays)

### ISSUE THREAD

**Issue by weinand** [Microsoft/vscode#9537]
Currently large data structures (Array, Map, Set) are broken into "chunks" in the debug adapter. Potentially this results in lots of duplication.

This feature request asks for doing this in the VS Code debugger UI once and for all.

An implementation sketch:

- if a variable, scope, or evaluate result has a `totalCount` attribute, check whether its value is larger than some (configurable) threshold.
- if yes then insert intermediate variable nodes between the parent and the children so that every node covers the range 0..threshold-1. Show this range as the label of the intermediate node.
- if an intermediate node is expanded, fetch only the corresponding range of children via the `variablesRequest` and the new `start` and `count` arguments.
- return a value of true for the `supportsVariablePaging` capability to indicate that the UI handles variable paging.

The corresponding protocol addition:  https://github.com/Microsoft/vscode-debugadapter-node/issues/31

**Comment by rozzzly** [comment]
Speaking of debugging Maps, Sets, etc.. actually viewing the values is huge a pain in the ass, you have to expand through like 4 levels to actually see it's data. Really time consuming when jumping from frame to frame!

How about something where when you expand the node of a Set/Map/etc object, its members are displayed as usual, but after those, there's a new "virtual* node which is collapsed. We could italicize, name it something like _`( expand to view data... )`_  to indicate the node doesn't actually exist on the . Expanding that node would display the entries within the Map/Set/etc.

I'm going to think more on this.. maybe submit a proposal.

**Comment by weinand** [comment]
@rozzzly when tare you talking about 4 level you are not talking about VS Code, right?
In VS Code you can see the value of an entry after expanding one level for small Maps (< 100) and two expands for any larger Maps:

![2016-07-21_08-16-09](https://cloud.githubusercontent.com/assets/1898161/17013059/f6b2ba76-4f1b-11e6-8138-4c899056041f.png)

![2016-07-21_08-24-02](https://cloud.githubusercontent.com/assets/1898161/17013124/84a228b2-4f1c-11e6-8a80-3696195b453e.png)

I don't understand how your virtual node would improve this.

**Comment by isidorn** [comment]
@weinand I have pushed the changes on the vscode side and I think it is in a good shape. I tried using it with the node adapter and I nicely get the `totalCount` attributes however when requesting a chunk of children I get back an empty response from the adapter. Thus forwarding this bug to you - to enable vscode to support paging just change the boolean [here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/debug/electron-browser/debugService.ts#L627)

**Comment by weinand** [comment]
@isidorn I have changed the property name to `totalVariables` as you suggested and I think we should enable this feature for July so that it can be adopted by debug adapters easily (I've already pushed this change so that it will become available in the Monday build).
However, I will not enable this feature in node-debug because we've already identified some limitations which would cripple node-debug.
To make testing possible, I've added an (unofficial) launch config flag `frontendVariablePaging` to enable variable paging in node-debug.
